coderd/workspaceupdates.go

```diff
@@ -0,0 +1,299 @@
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/pubsub"
+	"github.com/coder/coder/v2/coderd/util/slice"
+	"github.com/coder/coder/v2/coderd/wspubsub"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+)
+
+type workspacesByID = map[uuid.UUID]ownedWorkspace
+
+type ownedWorkspace struct {
+	WorkspaceName string
+	Status        proto.Workspace_Status
+	Agents        []database.AgentIDNamePair
+}
+
+// Equal does not compare agents
+func (w ownedWorkspace) Equal(other ownedWorkspace) bool {
+	return w.WorkspaceName == other.WorkspaceName &&
+		w.Status == other.Status
+}
+
+type sub struct {
+	mu     sync.RWMutex
+	userID uuid.UUID
+	tx     chan<- *proto.WorkspaceUpdate
+	prev   workspacesByID
+
+	db     UpdateQuerier
+	ps     pubsub.Pubsub
+	logger slog.Logger
+
+	cancelFn func()
+}
+
+func (s *sub) ownsAgent(agentID uuid.UUID) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, agent := range s.prev {
+		for _, a := range agent.Agents {
+			if a.ID == agentID {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (s *sub) handleEvent(ctx context.Context, event wspubsub.WorkspaceEvent) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	switch event.Kind {
+	case wspubsub.WorkspaceEventKindStateChange:
+	case wspubsub.WorkspaceEventKindAgentConnectionUpdate:
+	case wspubsub.WorkspaceEventKindAgentTimeout:
+	case wspubsub.WorkspaceEventKindAgentLifecycleUpdate:
+	default:
+		return
+	}
+
+	row, err := s.db.GetWorkspacesAndAgentsByOwnerID(context.Background(), s.userID)
+	if err != nil {
+		s.logger.Warn(ctx, "failed to get workspaces and agents by owner ID", slog.Error(err))
+	}
+	latest := convertRows(row)
+
+	out, updated := produceUpdate(s.prev, latest)
+	if !updated {
+		return
+	}
+
+	s.prev = latest
+	s.tx <- out
+}
+
+func (s *sub) start() (err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	rows, err := s.db.GetWorkspacesAndAgentsByOwnerID(context.Background(), s.userID)
+	if err != nil {
+		return xerrors.Errorf("get workspaces and agents by owner ID: %w", err)
+	}
+
+	latest := convertRows(rows)
+	initUpdate, _ := produceUpdate(workspacesByID{}, latest)
+	s.tx <- initUpdate
+	s.prev = latest
+
+	cancel, err := s.ps.Subscribe(wspubsub.WorkspaceEventChannel(s.userID), wspubsub.HandleWorkspaceEvent(s.logger, s.handleEvent))
+	if err != nil {
+		return xerrors.Errorf("subscribe to workspace event channel: %w", err)
+	}
+
+	s.cancelFn = cancel
+	return nil
+}
+
+func (s *sub) stop() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.cancelFn != nil {
+		s.cancelFn()
+	}
+
+	close(s.tx)
+}
+
+type UpdateQuerier interface {
+	GetWorkspacesAndAgentsByOwnerID(ctx context.Context, ownerID uuid.UUID) ([]database.GetWorkspacesAndAgentsByOwnerIDRow, error)
+}
+
+type updatesProvider struct {
+	mu sync.RWMutex
+	// Peer ID -> subscription
+	subs map[uuid.UUID]*sub
```

Comment:
This is keyed by Peer ID to quickly handle subscription deletions when disconnecting. We can't use user ID as it's not unreasonable to want to have multiple CoderVPN clients running at once.